lalrpop/src/rust/mod.rs

```diff
@@ -82,32 +92,48 @@ impl<'me, W: Write> RustWrite<W> {
                 if !first && session.emit_whitespace {
                     write!(self.write, " ")?;
                 }
-                write!(self.write, "{},", i)?;
+                write!(self.write, "{i},")?;
                 first = false;
             }
         }
         writeln!(self.write)
     }
 
-    pub fn writeln(&mut self, out: &str) -> io::Result<()> {
-        let buf = out.as_bytes();
+    // This function is intentionally not implemented with the `Write` trait
+    // because it is only called using the `rust!` macro, and thus with a `buf`
+    // that always ends with a `\n`.
+    #[doc(hidden)]
+    #[track_caller]
+    pub fn write_fmt(&mut self, args: fmt::Arguments<'_>) -> io::Result<()> {
+        // avoid intermediate allocation for simple format strings
+        let formatted;
+        let buf = match args.as_str() {
+            Some(s) => s.as_bytes(),
+            None => {
+                formatted = fmt::format(args);
+                formatted.as_bytes()
+            }
+        };
 
         // pass empty lines through with no indentation
-        if buf.is_empty() {
-            return self.write.write_all("\n".as_bytes());
+        match buf {
+            [] => return Ok(()),
+            [b'\n'] => return self.write.write_all(b"\n"),
+            [.., b'\n'] => {}
+            _ => unreachable!("rust! macro should always end with a newline"),
         }
 
-        let n = buf.len() - 1;
-
         // If the line begins with a `}`, `]`, or `)`, first decrement the indentation.
-        if buf[0] == b'}' || buf[0] == b']' || buf[0] == b')' {
+        if matches!(buf.first().unwrap(), b'}' | b']' | b')') {
             self.indent -= TAB;
         }
 
-        self.write_indented(out)?;
+        self.write_indentation()?;
+        self.write.write_all(buf)?;
 
-        // Detect a line that ends in a `{` or `(` and increase indentation for future lines.
-        if buf[n] == b'{' || buf[n] == b'[' || buf[n] == b'(' {
+        // If a line ends with a `{`, `[`, or `(`, increase indentation for future lines.
+        let n = buf.len().saturating_sub(2);
```

Comment:
I think the removed `n` calculation was `let n = buf.len() - 1;`, but now it seems to be `buf.len() - 2`?

Reply:
Ah, right, I see. Thanks.